rpcs3/Emu/Memory/vm.cpp

```diff
@@ -1,24 +1,324 @@
+	NamedThreadBase* g_reservation_owner = nullptr;
+
+	u32 g_reservation_addr = 0;
+
+	reservation_mutex_t g_reservation_mutex;
+
+	void _reservation_set(u32 addr)
+	{
+		//const auto stamp0 = get_time();
+
+#ifdef _WIN32
+		DWORD old;
+		if (!VirtualProtect(vm::get_ptr(addr & ~0xfff), 4096, PAGE_READONLY, &old))
+#else
+		if (mprotect(vm::get_ptr(addr & ~0xfff), 4096, PROT_READ))
+#endif
+		{
+			throw fmt::format("vm::_reservation_set() failed (addr=0x%x)", addr);
+		}
+
+		//LOG_NOTICE(MEMORY, "VirtualProtect: %f us", (get_time() - stamp0) / 80.f);
+	}
+
+	bool _reservation_break(u32 addr)
+	{
+		if (g_reservation_addr >> 12 == addr >> 12)
+		{
+			//const auto stamp0 = get_time();
+
+#ifdef _WIN32
+			if (!VirtualAlloc(vm::get_ptr(addr & ~0xfff), 4096, MEM_COMMIT, PAGE_READWRITE))
+#else
+			if (mprotect(vm::get_ptr(addr & ~0xfff), 4096, PROT_READ | PROT_WRITE))
+#endif
+			{
+				throw fmt::format("vm::_reservation_break() failed (addr=0x%x)", addr);
+			}
+
+			//LOG_NOTICE(MEMORY, "VirtualAlloc: %f us", (get_time() - stamp0) / 80.f);
+
+			if (g_reservation_cb)
+			{
+				g_reservation_cb();
+				g_reservation_cb = nullptr;
+			}
+
+			g_reservation_owner = nullptr;
+			g_reservation_addr = 0;
+
+			return true;
+		}
+
+		return false;
+	}
+
+	bool reservation_break(u32 addr)
+	{
+		std::lock_guard<reservation_mutex_t> lock(g_reservation_mutex);
+
+		return _reservation_break(addr);
+	}
+
+	bool reservation_acquire(void* data, u32 addr, u32 size, const std::function<void()>& callback)
+	{
+		const auto stamp0 = get_time();
+
+		bool broken = false;
+
+		assert(size == 1 || size == 2 || size == 4 || size == 8 || size == 128);
+		assert((addr + size & ~0xfff) == (addr & ~0xfff));
+
+		{
+			std::lock_guard<reservation_mutex_t> lock(g_reservation_mutex);
+
+			// silent unlocking to prevent priority boost for threads going to break reservation
+			g_reservation_mutex.do_notify = false;
+
+			// break previous reservation
+			if (g_reservation_addr)
+			{
+				broken = _reservation_break(g_reservation_addr);
+			}
+
+			// change memory protection to read-only
+			_reservation_set(addr);
+
+			// set additional information
+			g_reservation_addr = addr;
+			g_reservation_owner = GetCurrentNamedThread();
+			g_reservation_cb = callback;
+
+			// copy data
+			memcpy(data, vm::get_ptr(addr), size);
+		}
+
+		return broken;
+	}
+
+	bool reservation_update(u32 addr, const void* data, u32 size)
+	{
+		assert(size == 1 || size == 2 || size == 4 || size == 8 || size == 128);
+		assert((addr + size & ~0xfff) == (addr & ~0xfff));
+
+		std::lock_guard<reservation_mutex_t> lock(g_reservation_mutex);
+
+		if (g_reservation_addr != addr || g_reservation_owner != GetCurrentNamedThread())
+		{
+			// atomic update failed
+			return false;
+		}
 
-	/* OS X uses MAP_ANON instead of MAP_ANONYMOUS */
-	#ifndef MAP_ANONYMOUS
-	#define MAP_ANONYMOUS MAP_ANON
-	#endif
+		// update memory using privileged access
+		memcpy(vm::get_priv_ptr(addr), data, size);
```

Comment:
My understanding is that the reservation granule corresponds to a cache line and so it will always be updated atomically.

In any case I don't see how software (especially SPURS) can work reliably if the reservation cannot be acquired or updated atomically.